### ISSUE TITLE
out-sql performance

### ISSUE THREAD

**Issue by kakoni** [fluent/fluent-plugin-sql#39]
Trying to push data into psql database using out-sql plugin. 
I'm using in_tail plugin as a source (files with few million rows).

I've seen various error messages like
`failed to write data into buffer by buffer overflow`
and
`buffer space has too many data``

Any ideas where I should start with optimization? 
I tried with `buffer_queue_full_action block`but didnt help

**Comment by kakoni** [comment]
And also what sort of performance should be considered ok? (Well obviously depends on hardware). Seeing something like 300 imports / sec in current conf

**Comment by repeatedly** [comment]
> buffer space has too many data`

From this logs, do you use v0.14, right?
How about using v0.12 with `buffer_queue_full_action block`?

From my experience, out_sql is too slower than reading logs in in_tail.
If you want maximum performance for bulk loading, using embulk is better than fluentd.

I will check the performance on my Mac later.

**Comment by kakoni** [comment]
Yes, on 0.14. Cant go to 0.12, too much dependencies on 0.14;) (csv parsers etc..).

Few points
a) Are you saying `buffer_queue_full_action` doesnt work with 0.14?
b) embulk. Well yes and no. fluent has position files etc, all the features I need for my continuous batch jobs, something (that I believe) is missing from embulk?

**Comment by kakoni** [comment]
Quick note, made progress with
` 
num_threads 5
flush_interval 1s
buffer_queue_full_action block`
(on 0.14). Comment?

**Comment by Watson1978** [comment]
Thanks for the report, and apologies for the very long silence on this issue.

This is a performance question rather than a bug. 
The behavior you saw comes from a design characteristic of this plugin: `out_sql` writes records through ActiveRecord, so it is not well suited to bulk-loading millions of rows. 
When in_tail reads far faster than `out_sql` can insert, the buffer fills up and you get `buffer overflow` / `buffer space has too many data`.

A few notes:

- As mentioned earlier in this thread, if your goal is high-throughput bulk loading, a dedicated tool such as [embulk](https://www.embulk.org/) is a much better fit than fluentd + `out_sql`.
- If you do stay on fluentd, throughput can be improved by increasing flush concurrency and tuning the buffer. Note that the config syntax has changed significantly since the fluentd 0.12/0.14 era this issue is from. On current fluentd it looks roughly like this:

```
  <match ...>
    @type sql
    # ... connection / table settings ...
    <buffer>
      flush_thread_count 8
      flush_interval 1s
      chunk_limit_size 8m
      total_limit_size 512m
      overflow_action block
    </buffer>
  </match>
```

`overflow_action block` applies back-pressure so input slows down instead of dropping data, at the cost of throughput.

Since this issue is based on a very old version and the underlying cause is a known limitation rather than a bug, I'm going to close it. If you hit a buffering problem on a current version of fluentd and this plugin, please open a new issue with your full config and version numbers and we'll take a look.
